x/multisig/msg.go

```diff
@@ -44,19 +36,62 @@ func (UpdateContractMsg) Path() string {
 
 // Validate enforces sigs and threshold boundaries
 func (c *UpdateContractMsg) Validate() error {
-	if len(c.Sigs) == 0 {
-		return errors.Wrap(errors.ErrInvalidMsg, "missing sigs")
+	switch n := len(c.Participants); {
+	case n == 0:
+		return errors.Wrap(errors.ErrInvalidMsg, "no participants")
+	case n > 100:
+		return errors.Wrap(errors.ErrInvalidMsg, "too many participants")
 	}
-	if c.ActivationThreshold <= 0 || int(c.ActivationThreshold) > len(c.Sigs) {
-		return errors.Wrap(errors.ErrInvalidMsg, invalidThreshold)
-	}
-	if c.AdminThreshold <= 0 {
-		return errors.Wrap(errors.ErrInvalidMsg, invalidThreshold)
+	return validateWeights(errors.ErrInvalidMsg,
+		c.Participants, c.ActivationThreshold, c.AdminThreshold)
+}
+
+// validateWeights returns an error if given participants and thresholds
+// configuration is not valid. This check is done on model and messages so
+// instead of copying the code it is extracted into this function.
+func validateWeights(
+	baseErr error,
```

Comment:
Interesting that baseErr is only used for two of the cases (admin < activation and activation > total).
The others return same error from Validate.
Is this proper? (this is not a rhetorical question, I just want you to think about it and answer)

Reply:
This is intended. If I am getting an error, I want to preserve it, so that it builds a nice chain of failures. Only if the condition is invalid and I have to create an error from scratch then I need some base error.
This function is used by both message and model validation so the base error differs.

I generally don't like those errors (`ErrInvalidMsg` and `ErrInvalidModel`) and I think we would be much better with a more generic `ErrInvalid` that covers all validation issues. Any detail will be part of the error message.